src/services/completions.ts

```diff
@@ -335,6 +322,56 @@ namespace ts.Completions {
         }
     }
 
+    function getJsxClosingTagCompletion(location: Node | undefined, sourceFile: SourceFile): CompletionInfo | undefined {
+        // When there is a full closing tag and the tag includes a property access expression,
+        // the node hierarchy to get the JSX closing element is at grand parent of location.
+        // For example:
+        //   var x = <Comp.Item></Comp.Item/*1*/
+        //   var y = <Comp.Item></Comp.Item/*2*/>
+        const jsxClosingElement = location && location.parent
+            ? location.parent.parent && isJsxClosingElement(location.parent.parent)
+                ? location.parent.parent
+                : isJsxClosingElement(location.parent)
+                    ? location.parent
+                    : undefined
+            : undefined;
+
+        if (jsxClosingElement) {
+            // In the TypeScript JSX element, if such element is not defined. When users query for completion at closing tag,
+            // instead of simply giving unknown value, the completion will return the tag-name of an associated opening-element.
+            // For example:
+            //     var x = <div> </ /*1*/
+            // The completion list at "1" will contain "div>" with type any
+            // And at `<div> </ /*1*/ >` (with a closing `>`), the completion list will contain "div".
+            // And at property access expressions `<MainComponent.Child> </MainComponent. /*1*/ >` the completion will
+            // return full closing tag with an optional replacement span
+            // For example:
+            //     var x = <MainComponent.Child> </     MainComponent /*1*/  >
+            //     var y = <MainComponent.Child> </   /*2*/   MainComponent >
+            // the completion list at "1" and "2" will contain "MainComponent.Child" with a replacement span from `/` (slash) till `>`
+            // resulting in a clean closing tag
+            const hasClosingAngleBracket = !!findChildOfKind(jsxClosingElement, SyntaxKind.GreaterThanToken, sourceFile);
+            const slashToken = findChildOfKind(jsxClosingElement, SyntaxKind.SlashToken, sourceFile);
+            const replacementPos = slashToken ? slashToken.pos +1 : jsxClosingElement.pos +2;
+            const replacementEnd = hasClosingAngleBracket ? jsxClosingElement.end -1 : jsxClosingElement.end;
+            const tagName = jsxClosingElement.parent.openingElement.tagName;
+            const closingTag = tagName.getText(sourceFile);
+            const fullClosingTag = closingTag + (hasClosingAngleBracket ? "" : ">");
+            const replacementSpan = replacementEnd - replacementPos > 0
+                ? createTextSpanFromRange({ pos: replacementPos, end: replacementEnd })
+                : undefined;

```

Comment:
```suggestion
            const replacementSpan = createTextSpanFromNode(jsxClosingElement.tagName);
```

and delete `replacementPos`/`replacementEnd`

Reply:
If we use `jsxClosingElement.tagName` to get the text span it won't select the whitespaces after slash. Is this okay?

my implementation would do `</    Exp. >` -> `</Exp.M>`

Reply:
Yeah, we don’t want to reformat whitespace. The completion isn’t supposed to feel like it’s replacing/reformatting anything.

Reply:
Done.